api/v1alpha3/azureimage_types.go

```diff
@@ -0,0 +1,109 @@
+// +kubebuilder:resource:path=azureimages,scope=Namespaced,categories=cluster-api
+// +kubebuilder:storageversion
+
+// AzureMarketplaceImage defines an image in the Azure marketplace to use for VM creation
+type AzureMarketplaceImage struct {
+	metav1.TypeMeta `json:",inline"`
+
+	// Publisher is the name of the organization that created the image
+	// +kubebuilder:validation:MinLength=1
+	Publisher string `json:"publisher"`
+	// Offer specifies the name of a group of related images created by the publisher.
+	// For example, UbuntuServer, WindowsServer
+	// +kubebuilder:validation:MinLength=1
+	Offer string `json:"offer"`
+	// SKU specifies an instance of an offer, such as a major release of a distribution.
+	// For example, 18.04-LTS, 2019-Datacenter
+	// +kubebuilder:validation:MinLength=1
+	SKU string `json:"sku"`
+	// Version specifies the version of an image sku. The allowed formats
+	// are Major.Minor.Build or 'latest'. Major, Minor, and Build are decimal numbers.
+	// Specify 'latest' to use the latest version of an image available at deploy time.
+	// Even if you use 'latest', the VM image will not automatically update after deploy
+	// time even if a new version becomes available.
+	// +kubebuilder:validation:MinLength=1
+	Version string `json:"version"`
+}
+
+// +kubebuilder:object:root=true
+// +kubebuilder:resource:path=azureimages,scope=Namespaced,categories=cluster-api
+// +kubebuilder:storageversion
+
+// AzureSharedGalleryImage defines an image in a Shared Image Gallery to use for VM creation
+type AzureSharedGalleryImage struct {
+	metav1.TypeMeta `json:",inline"`
+
+	// SubscriptionID is the identifier of the subscription that contains the shared image gallery
+	// +kubebuilder:validation:MinLength=1
+	SubscriptionID string `json:"subscriptionID"`
+	// ResourceGroup specifies the resource group containing the shared image gallery
+	// +kubebuilder:validation:MinLength=1
+	ResourceGroup string `json:"resourceGroup"`
+	// Gallery specifies the name of the shared image gallery that contains the image
+	// +kubebuilder:validation:MinLength=1
+	Gallery string `json:"gallery"`
+	// Name is the name of the image
+	// +kubebuilder:validation:MinLength=1
+	Name string `json:"name"`
+	// Version specifies the version of the marketplace image. The allowed formats
+	// are Major.Minor.Build or 'latest'. Major, Minor, and Build are decimal numbers.
+	// Specify 'latest' to use the latest version of an image available at deploy time.
+	// Even if you use 'latest', the VM image will not automatically update after deploy
+	// time even if a new version becomes available.
+	// +kubebuilder:validation:MinLength=1
+	Version string `json:"version"`
+}
+
+// +kubebuilder:object:root=true
+// +kubebuilder:resource:path=azureimages,scope=Namespaced,categories=cluster-api
+// +kubebuilder:storageversion
+
+// AzureImageByID defines a specific image via by resource id to use for VM creation
+type AzureImageByID struct {
```

Comment:
nit: would be nice if the naming style matched (AzureXXXImage), but I can't come up with anything better.

Reply:
+1

If we end up changing to https://github.com/kubernetes-sigs/cluster-api-provider-azure/issues/293#issuecomment-594226231, this can just become `ID`.